### PR TITLE
Functionality for loading comments from most recent to least

### PR DESCRIPTION
**Pull request by Anthony-Fiddes**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -14,19 +14,56 @@
...
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  static final String CONTENT_TYPE = "application/json;";
```
Comment (Alfredhgoog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:36]: Remove semicolon from string

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -14,19 +14,56 @@
...
+    String json = gson.toJson(messages);
+    response.setContentType(CONTENT_TYPE);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String content = request.getParameter("content").trim();
+    long timestamp = System.currentTimeMillis();
+    if (content.length() == 0) {
```
Comment (Alfredhgoog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:58]: Nit - Prefer String.isEmpty

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -14,19 +14,56 @@
...
+      System.err.println("Empty comment submitted");
+      return;
```
Comment (Alfredhgoog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:60]: What happens if you don't specify a response? Would the user be redirected to a blank page instead?

Consider adding a default HttpServletResponse.sendError, or you may (consider it a separate feature for error handling) redirect user to a customized error handling page, such as this one: https://github.com/thisgoesnowhere

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,19 +14,56 @@
...
+    }
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("content", content);
+    commentEntity.setProperty("timestamp", timestamp);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
```
Comment (Alfredhgoog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:66]: Not an immediate actionable item requested, but datastore service might be unavailable due to network conditions, service linkage etc. A failure will result in the servlet throwing an exception and showing a stacktrace to the response webpage, not so good looking as a web browsing experience. I would do a try {} catch {} for the specific exceptions, and return a friendly error message.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (Anthony-Fiddes) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:60]: So I ended up realizing this later in my coding journey. In this PR, they are directed to a blank page, but commit e0f906f3feb09751c94b813a9a4b6ed8c433f860 will redirect the user no matter what. Is there a way to let the user know that they were unsuccessful, while still remaining at the comment box?

Comment (Alfredhgoog) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:64]: Great question.

"know that they were unsuccessful, while still remaining at the comment box"

Form submission *is* a redirection so user will land on whatever you response.redirect(...) them to. In order to keep the page from being navigated, you need to send the form in JavaScript. AJAX is the basic format of how this is done. jQuery is a library to simplify this process:
https://api.jquery.com/jquery.ajax/

It is easy to ramp up, and may lend great help for the Capstone project as well.

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -16,7 +16,13 @@
  * Adds comments from /data
  */
 async function getComments() {
-    const commentsData = await fetch("/data");
-    const comments = await commentsData.text();
-    document.getElementById("comments-container").innerHTML = comments;
+  const commentsData = await fetch('/data');
+  const comments = await commentsData.json();
+  const ul = document.createElement('ul');
+  for (let i = 0; i < comments.length; i++) {
```
Comment (Alfredhgoog) [inline portfolio/src/main/webapp/script.js:22]: Prefer ES6 style
for (let [comment, index] of comments) {...}

The square brackets are called destructuring.
https://exploringjs.com/es6/ch_for-of.html

**Review by Alfredhgoog — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (Anthony-Fiddes) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:70]: Opened an issue: https://github.com/Anthony-Fiddes/portfolio/issues/13

**Review by ricazhang — APPROVED** [review]

**Pull request merged**